Datagrid docs

mdvictor opened the pull request:
Docs for the upcoming experimental datagrid panel
Images uploaded here: https://admin.grafana.com/upload/media/docs/datagrid/

img urls:
https://grafana.com/media/docs/datagrid/screenshot-grafana-datagrid-panel.png
https://grafana.com/media/docs/datagrid/screenshot-grafana-datagrid-context-menu.png
https://grafana.com/media/docs/datagrid/screenshot-grafana-datagrid-header-menu.png

mdvictor committed:
WIP Datagrid docs

mdvictor committed:
Finish datagrid docs

mdvictor committed:
Add experimental note

mdvictor committed:
Fix typo

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+> **Note:** Grafana Datagrid panel is Experimental. This feature is supported by the engineering team on a best-effort basis, and breaking changes may occur without notice prior to general availability.
+
+The datagrid panel offers a user the ability to create or edit and fine-tune data within Grafana. As such, this panel can act as a data source for other panels
```
```suggestion
The datagrid panel offers a user the ability to create, edit, and fine-tune data within Grafana. As such, this panel can act as a data source for other panels
```

baldm0mma replied:
Also, should `datagrid` be capitalized?

imatwawana replied:
I would agree that Datagrid should be capitalized here and throughout to conform with the way we currently tend to document panels (this might change, but for now, let's remain consistent).

mdvictor replied:
Thanks for this. I never knew you can add a comma before `and` when listing qualities. I always thought it's a mistake to add it in that case, but apparently based on [this](https://www.grammarly.com/blog/comma-before-and/) it's optional.

baldm0mma replied:
Yeah, I don't have **_super_** strong feelings about comma usage, but added it here since the Oxford usage is appropriate for the syntactic context.

imatwawana replied:
Our style guidelines for Grafana are to use Oxford or serial commas, so definitely add it.

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+## Use cases
+
+The datagrid panel can be used to manipulate data queried from any data source or from a dragged and dropped file. It can then be used as a simple tabular
```
```suggestion
The datagrid panel can be used to manipulate data queried from any data source, or from a dragged and dropped file. It can then be used as a simple tabular
```

mdvictor replied:
I'm curious about the rules here. I would normally write without a comma, is it because the second part after the `or`is an independent clause? Would writing it this way, without the `or` be correct? - `queried from any data source or dragged and dropped file`

baldm0mma replied:
The comma here indicates a pause, which I think clarifies the distinctions between the 2 options ("any data source" || "drag and dropped file"). But, again, no strong opinion either way; I think both with and without the comma are correct.

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+## Use cases
+
+The datagrid panel can be used to manipulate data queried from any data source or from a dragged and dropped file. It can then be used as a simple tabular
+visualization or the data can be modified and even removed altogether to create a blank slate. Any changes made to data source retrieved data trigger a confirmation modal that informs the user about the action they are about to take. Editing the dataset changes the data source to use the inbuilt `-- Grafana --` data source, thus losing the old data source settings and related queries, while also  ...
```
```suggestion
visualization, or the data can be modified - and even removed altogether - to create a blank slate. Any changes made to data source retrieved data triggers a confirmation modal that informs the user about the action they are about to take. Editing the dataset changes the data source to use the inbuilt `-- Grafana --` data source, thus losing the old data source settings and related queries, while also copying the current dataset into the dashboards model. After confirming the action, all data modifications are saved to the dashboard model.
```

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+## Use cases
+
+The datagrid panel can be used to manipulate data queried from any data source or from a dragged and dropped file. It can then be used as a simple tabular
+visualization or the data can be modified and even removed altogether to create a blank slate. Any changes made to data source retrieved data trigger a confirmation modal that informs the user about the action they are about to take. Editing the dataset changes the data source to use the inbuilt `-- Grafana --` data source, thus losing the old data source settings and related queries, while also  ...
```
Is there documentation about the `-- Dashboard --` data source that you should link here?

mdvictor replied:
That is a good question, I have not found any docs related to the in-built `Dashboard` or `Grafana` datasources. @imatwawana do you know of any?

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+### Adding data
+
+Data can be added to the datagrid by creating a new column or row. To create a new column, click the `+` button found after the last column. After clicking the button, a new input will prompt the user to name the newly created column. Clicking anywhere outside the input or pressing the `Enter` key will finalize column creation. After that, data can be filled on each cell.
```
```suggestion
Data can be added to the datagrid by creating a new column or row. To create a new column, click the `+` button found after the last column. After clicking the button, a new input will prompt the user to name the newly created column. Clicking anywhere outside the input, or pressing the `Enter` key, will finalize column creation. After that, data can be filled on each cell.
```

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+### Selecting multiple cells
+
+A user can select multiple cells by clicking on one and dragging the mouse across others. This selection can be used to copy the data from the selected cells or to delete them using the `Delete` key.
```
```suggestion
A user can select multiple cells by clicking on a single cell and dragging the mouse across others. This selection can be used to copy the data from the selected cells or to delete them using the `Delete` key.
```

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+### Using the context menu
+
+To provide a more streamlined experience, the datagrid panel offers a context menu that can be accessed on right click. Based on the state of the datagrid, the context menu will offer different options. The context menu can be accessed by right clicking on a cell, column header or row selector. Through the context menu, users can delete or clear rows and column, remove all existing data (renderin ...
```
```suggestion
To provide a more streamlined experience, the datagrid panel offers a context menu that can be accessed on right mouse click. Based on the state of the datagrid, the context menu will offer different options. The context menu can be accessed by right mouse clicking on a cell, column header, or row selector. Through the context menu, users can delete or clear rows and columns, remove all existing data (rendering the datagrid blank), or trigger the search functionality. Deleting a row or column will remove the data from the datagrid, while clearing a row or column will only remove the data from the cells, leaving the row or column intact.
```

baldm0mma commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+### Header dropdown menu
+
+There is another type of menu dedicated to the headers. Clicking on the dropdown icon on the right side of the header title will open a new menu from which users can also delete or clear a column, but also rename it, freeze it or convert the field type of the column.
```
```suggestion
There is another type of menu dedicated to the headers. Clicking on the dropdown icon on the right side of the header title will open a new menu from which users can also delete or clear a column, but also rename it, freeze it, or convert the field type of the column.
```

baldm0mma reviewed (approved):
This is great! A few nits.

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+  - datagrid options
+menuTitle: Datagrid
+title: Datagrid
+weight: 950
```
Moves this into alphabetical order
```suggestion
weight: 350
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+# Datagrid
+
+> **Note:** Grafana Datagrid panel is Experimental. This feature is supported by the engineering team on a best-effort basis, and breaking changes may occur without notice prior to general availability.
```
Formatting note using a shortcode 
```suggestion
{{% admonition type="note" %}}

The Grafana Datagrid panel is Experimental. This feature is supported by the engineering team on a best-effort basis, and breaking changes may occur without notice prior to general availability.

{{% /admonition %}}
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,82 @@
+
+> **Note:** Grafana Datagrid panel is Experimental. This feature is supported by the engineering team on a best-effort basis, and breaking changes may occur without notice prior to general availability.
+
+The datagrid panel offers a user the ability to create or edit and fine-tune data within Grafana. As such, this panel can act as a data source for other panels
```
```suggestion
The Datagrid panel offers you the ability to create, edit, and fine-tune data within Grafana. As such, this panel can act as a data source for other panels
```

imatwawana reviewed (commented):
I started a review, but given that @baldm0mma's review was focused on wording, I think you should implement those changes first and then I can do a further review after that. It's difficult to do a review focused on the same thing on top of another one. Please ping me for review when you're ready for me to look again.

mdvictor committed:
PR modifications for datagrid docs

mdvictor committed:
docs modifications

mdvictor commented:
Thank you @baldm0mma and @imatwawana for your suggestions. I made the modifications accordingly.

I've also added a small section referencing the Special datasources docs section and changed the doc to refer to `you` instead of `a users/ the user/ users`. I am looking forward for anything else you might find! Thank you!

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+### Using the context menu
+
+To provide a more streamlined experience, the datagrid panel offers a context menu that can be accessed on right mouse click. Based on the state of the datagrid, the context menu will offer different options. The context menu can be accessed by right mouse clicking on a cell, column header, or row selector. Through the context menu, you can delete or clear rows and columns, remove all existing da ...
```
```suggestion
To provide a more streamlined experience, the Datagrid panel has a context menu that can be accessed by right clicking on a cell, column header, or row selector. Depending on the state of your datagrid, the context menu offers different options including:

- Delete or clear rows and columns.
- Remove all existing data (rendering your datagrid blank).
- Trigger search functionality, which allows you to find keywords within the dataset. 

Deleting a row or column will remove the data from the datagrid, while clearing a row or column will only remove the data from the cells, leaving the row or column intact.

### Header menu

You can also access a header menu by clicking the dropdown icon next to the header title. From here, you can not only delete or clear a column, but also rename it, freeze it, or convert the field type of the column.

![Datagrid header menu](/media/docs/datagrid/screenshot-grafana-datagrid-header-menu.png)
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+You can select multiple cells by clicking on a single cell and dragging the mouse across others. This selection can be used to copy the data from the selected cells or to delete them using the `Delete` key.
+
+### Using the context menu
```
This section is more descriptive of a UI element than really directions on how to use a context menu—folks know how to use a context menu. I would move this up to just before the use case section and make it an H2. I've also suggested moving the header menu content under this since they're both menus.
```suggestion
### Context menu
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+To delete all rows, use the `Select all` checkbox found on the top left corner of the datagrid. This will select all rows and allow the user to delete them through the context menu.
+
+### Searching data
+
+The search functionality can be triggered from the context menu and will allow you to find keywords within the dataset. The searchbox can then be closed either by clicking the `x` button or pressing the `Escape` key.
+
```
The second sentence seems like overkill to me, which then renders this a one-sentence section, which isn't ideal. I'd kill this section and I've made the first sentence part of the search bullet point in the context menu section.
```suggestion
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+The search functionality can be triggered from the context menu and will allow you to find keywords within the dataset. The searchbox can then be closed either by clicking the `x` button or pressing the `Escape` key.
+
+### Header dropdown menu
+
+There is another type of menu dedicated to the headers. Clicking on the dropdown icon on the right side of the header title will open a new menu from which users can also delete or clear a column, but also rename it, freeze it, or convert the field type of the column.
+Columns can be converted between `numeric`, `string` and `boolean` types.
+
+![Datagrid header menu](/media/docs/datagrid/screenshot-grafana-datagrid-header-menu.png)
```
```suggestion
```

imatwawana reviewed (changes requested):
This is a really good start! I think it needs some streamlining and restructuring to make sure we're providing users with content that helps them achieve specific goals vs "here's everything the feature can do", if that makes sense. I've made a number of wording edits and some structural edits and left suggestions for some other structural and formatting changes that I think will improve the content. 

Also, in terms of where to capitalize/not capitalize Datagrid, if you're talking about it as an entire feature then capitalize and if you're talking about a specific instance of a datagrid that a user has created, then don't. I've tried to implement that in my edits, but I may not have caught every instance.

mdvictor commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+## Use cases
+
+The datagrid panel can be used to manipulate data queried from any data source, or from a dragged and dropped file. It can then be used as a simple tabular
+visualization, or the data can be modified - and even removed altogether - to create a blank slate. Any changes made to data source retrieved data triggers a confirmation modal that informs you about the action you are about to take. Editing the dataset changes the data source to use the inbuilt `-- Grafana --` data source, thus losing the old data source settings and related queries, while also  ...
```
Hey @imatwawana.

I agree we don't need to give info regarding 1 and 4. For 2, we're changing the datasources because a typical workflow would involve pulling some data from some datasource (e.g: mysql) and then the user might want to fine-tune the resulted data. In that case, the entire dataset will be saved as a snapshot which is a query type under the `-- Grafana --` datasource. This is to show that we no long query data from MySQL, but use locally stored data. For 3 I also agree that `replace` is better.

imatwawana replied:
Okay i've updated my edit to incorporate removing the stuff from 1 and 4 and hopefully accurately reflecting your explanation.

mdvictor replied:
Thank you! I just want to mention that it does not have to be MySQL neccesarily, but any datasource can be used to pull data and then editing it through the datagrid will copy and move that data to a Grafana DS Snapshot

mdvictor commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+![Datagrid panel](/media/docs/datagrid/screenshot-grafana-datagrid-panel.png)
+
+## Use cases
```
Yeah, this section might be a revamp. I guess there is only one use-case with some minor variations really: Pull data from somewhere -> fine-tune -> visualize in other panels vs Create dataset from scratch -> visualize it. Or even just pull data and use the datagrid as a simple, less formatted, table viz.

imatwawana replied:
Okay i get what you mean and having a heading that just says "use case" feels a bit off. one way you might get around that is to describe the one use case but then provide a couple example scenarios?

mdvictor commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,88 @@
+
+### Adding data
+
+Data can be added to the datagrid by creating a new column or row. To create a new column, click the `+` button found after the last column. After clicking the button, a new input will prompt you to name the newly created column. Clicking anywhere outside the input, or pressing the `Enter` key, will finalize column creation. After that, data can be filled on each cell.
+
+To add a new row, click the `+` button found after the last row. The button is present on each cell after the last row and will trigger creating a new one while also activating the corresponding cell that was clicked.
```
To use the panel, you need to get data from any datasource and as soon as you edit it (by changing cells or adding or removing cols/rows) it changes to `Grafana` DS with snapshot query type.

mdvictor committed:
PR modifications

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid-docs

mdvictor commented:
Hey @imatwawana! I modified the docs as per your suggestions. Looking forward for your input on the docs. Thank you!

baldm0mma commented:
@mdvictor doc-validator's failing, but a merge from main should take care of it. Thanks @imatwawana and @Eve832 !

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,117 @@
+
+You can also access a header menu by clicking the dropdown icon next to the header title. From here, you can not only delete or clear a column, but also rename it, freeze it, or convert the field type of the column.
+
+![Datagrid header menu](/media/docs/datagrid/screenshot-grafana-datagrid-header-menu.png)
```
This is another image where the point of the image gets a bit lost. I'd suggest just cropping out the first three rows of the grid to give the context menu more focus. I know we'll lose the name of the panel, but that's fine.

imatwawana reviewed (commented):
This is looking great! I like how you updated the use cases section—I think that makes a lot of sense. I've gone through and made mostly styling and wording edits at this point and moved one screenshot. I've also left suggestions about cropping a couple screenshots to make the focus more clear. I'm fairly certain this will be my last set of feedback. When you've implemented these changes, just ping me for one more quick look so I can make sure I haven't missed anything. Thanks for all your work on improving this!! Much appreciated!

mdvictor commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,117 @@
+Through it, you can manipulate data queried from any data source, or from a dragged and dropped file. You can then use the panel as a simple tabular
+visualization, or you can modify the data—and even remove it altogether—to create a blank slate.
+
+Editing the dataset changes the data source to use the inbuilt `-- Grafana --` data source, thus replacing the old data source settings and related queries, while also copying the current dataset into the dashboard model. You can either use a pre-existing dataset from any other data source, start with a blank slate, or you can drag and drop a file to create a new dataset.
```
I'll be changing the above first sentence to `Through it, you can manipulate data queried from any data source, you can start from a blank slate, or you can pull data from a dragged and dropped file.` so that it encapsulates the details and remove the last sentence of the second part alltogether.

mdvictor commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,117 @@
+
+1. In an existing panel, click the **+** button after the last column.
+1. When prompted, add a name for the new column.
+1. Click anywhere outside the field or press the `Enter` key to save the column.
```
Removing all code formats from all keys (`Delete`, `Enter`, `Cmd`, `Shift`, `Ctrl`)

mdvictor committed:
PR modifications

mdvictor committed:
PR mods

mdvictor commented:
@imatwawana Thank you so much for all the help! I really appreciate it! I made the changes and pushed a new version of the docs. Do tell me if there's anything else that needs to be added.

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+You can edit data by taking the following steps:
+
+1. Double click on the cell that needs to be modified. This will activate the cell and allow you to edit the data.
```
```suggestion
1. Double-click on the cell that needs to be modified. This will activate the cell and allow you to edit the data.
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+1. Drag the column to the desired location.
+1. Release the mouse button to finalize the move.
+
+To move a row, click and hold on the row selector from the number column situated on the left side, and drag it to the desired location. Releasing the mouse button finalizes the move.
```
```suggestion
To move a row, click and hold on the row selector from the number column situated on the far left side of the grid, and drag it to the desired location. Releasing the mouse button finalizes the move.
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+To delete or clear multiple rows, take the following steps:
+
+1. Hover over the number column (to the left of the first column in the grid) to display its column checkbox.
```
```suggestion
1. Hover over the number column (to the left of the first column in the grid) to display row checkbox.
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+1. Hover over the number column (to the left of the first column in the grid) to display its column checkbox.
+1. Select the checkboxes for the rows you want to work with.
+   To select multiple consecutive rows, you must hold the Shift key pressed while clicking on the first and last row. To select certain rows, the Ctrl (or Cmd) key must be pressed while clicking on the desired rows.
```
```suggestion
   To select multiple consecutive rows, press and hold the Shift key while clicking on the first and last row. To select certain rows, press and hold the Ctrl (or Cmd) key while clicking the desired rows.
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+1. Right-click to access the context menu.
+1. Select **Delete rows** or **Clear rows**.
+
+The same rules apply to columns by clicking on the column headers.
```
```suggestion
The same rules apply to columns by clicking the column headers.
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+Deleting a row or column will remove the data from the datagrid, while clearing a row or column will only remove the data from the cells, leaving the row or column intact.
+
+![Datagrid context menu](/media/docs/datagrid/screenshot-grafana-datagrid-context-menu-2.png)
```
```suggestion
{{< figure src="/media/docs/datagrid/screenshot-grafana-datagrid-context-menu-2.png" alt="Datagrid context menu" max-width="400px" >}}
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+![Datagrid context menu](/media/docs/datagrid/screenshot-grafana-datagrid-context-menu-2.png)
+
+### Header menu
```
```suggestion
## Header menu
```

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
```diff
@@ -0,0 +1,118 @@
+
+You can also access a header menu by clicking the dropdown icon next to the header title. From here, you can not only delete or clear a column, but also rename it, freeze it, or convert the field type of the column.
+
+![Datagrid header menu](/media/docs/datagrid/screenshot-grafana-datagrid-header-menu-2.png)
```
```suggestion
{{< figure src="/media/docs/datagrid/screenshot-grafana-datagrid-header-menu-2.png" alt="Datagrid header menu" max-width="500px" >}}```

imatwawana committed:
missed wording edits; added max size to screenshots; changed heading level

imatwawana commented on docs/sources/panels-visualizations/visualizations/datagrid/index.md:
````diff
@@ -0,0 +1,118 @@
+
+You can also access a header menu by clicking the dropdown icon next to the header title. From here, you can not only delete or clear a column, but also rename it, freeze it, or convert the field type of the column.
+
+{{< figure src="/media/docs/datagrid/screenshot-grafana-datagrid-header-menu-2.png" alt="Datagrid header menu" max-width="500px" >}}```
````
```suggestion
{{< figure src="/media/docs/datagrid/screenshot-grafana-datagrid-header-menu-2.png" alt="Datagrid header menu" max-width="500px" >}}
```

imatwawana committed:
fix typo

imatwawana committed:
style fix and change heading level

imatwawana reviewed (approved):
I made some last wording edits that I missed previously, limited the width of the updated screenshots (thank you for that) so they look a bit more balanced with the main screenshot, and changed the Header menu level back to an H2—my suggestion to make it an H3 doesn't make sense in the new page structure when you're looking at the page TOC. This looks great!
